packages/job-manager/src/batch/pool-metrics.ts

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { injectable } from 'inversify';
+
+export interface PoolLoad {
+    activeTasks: number;
+    runningTasks: number;
+    pendingTasks: number;
+}
+
+export interface PoolMetricsInfo {
+    id: string;
+    maxTasksPerPool: number;
+    load: PoolLoad;
+}
+
+export interface PoolMetricsState {
+    lastPoolLoad?: PoolLoad;
+    lastTasksIncrementCount?: number;
+    processingSpeed?: number;
+}
+
+@injectable()
+export class PoolMetrics {
+    public readonly poolState: PoolMetricsState = {};
+
+    public getTasksIncrementCount(poolMetricsInfo: PoolMetricsInfo, targetQueuedTasksOverloadRatio: number): number {
+        // No last pool state available. Use default task increment count.
+        if (this.poolState.lastTasksIncrementCount === undefined) {
+            this.poolState.lastPoolLoad = poolMetricsInfo.load;
+            this.poolState.processingSpeed = 0;
+            this.poolState.lastTasksIncrementCount = poolMetricsInfo.maxTasksPerPool * targetQueuedTasksOverloadRatio;
+
+            return this.poolState.lastTasksIncrementCount;
+        }
+
+        // Calculate processing speed since last tasks increment.
+        this.poolState.processingSpeed =
+            this.poolState.lastTasksIncrementCount - (poolMetricsInfo.load.activeTasks - this.poolState.lastPoolLoad.activeTasks);
```

Comment:
Is it possible that the processing speed is very fast so that in each sample there is 0 active tasks?

Reply:
Or if the processing speed is very slow, then will the activeTasks will always be maxTasksPerPool?

Reply:
- yes it is theoretically possible
- no activeTasks can be more than maxTasksPerPool